src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -149,13 +150,22 @@ public void Dispose()
 
         private void Reset(bool isDisposing = false)
         {
+            // Make sure worker restore operation is cancelled
             _workerCts?.Cancel();
 
             if (_backgroundJobRunner?.IsValueCreated == true)
             {
-                // Do not block VS for more than 5 sec.
+                // Await completion of the background work
                 _joinableFactory.Run(
-                    () => Task.WhenAny(_backgroundJobRunner.Value, Task.Delay(TimeSpan.FromSeconds(5))));
+                    async () =>
+                    {
+                        using (_joinableCollection.Join())
+                        {
+                            // Do not block VS forever
+                            await Task.WhenAny(_backgroundJobRunner.Value, Task.Delay(TimeSpan.FromSeconds(60)));
```

Comment:
Several issues here:

 * I see `_backgroundJobRunner` is typed as  `Lazy<Task>`, which is (generally speaking) a very dangerous type, and one we now have [a threading analyzer](https://github.com/Microsoft/vs-threading/blob/v15.0/doc/analyzers/VSTHRD011.md) to guard against. In this case you'd be OK if that Task belongs to the `_joinableCollection` since you join it, but I don't see you doing that either at least where you construct the Task. Use `AsyncLazy<T>` instead of `Lazy<T>`, and pass it your JTF instance to guard against multiple causes of deadlocks.
* Awaiting a Task within JTF.Run will deadlock if anything in that async Task needed the UI thread (directly or indirectly). As you've joined a collection here, you guard against deadlocks _if_ everything in that `Task` is a member of that collection, but I can't tell clearly whether that's the case. Again, switching to `AsyncLazy<T>` will guard against this. Alternatively you can safely await `JoinableTask` within a JTF.Run (assuming you're following the threading rules elsewhere). But generally speaking the only `Task` that is safe to await in a JTF.Run delegate is one you *created in that same delegate*.